packages/yarnpkg-core/sources/Project.ts

```diff
@@ -2015,6 +2060,172 @@ function applyVirtualResolutionMutations({
+          subject: dependentHash,
+          requested: ident,
+          rootRequester: rootHash,
+          allRequesters: linkHashes,
+        });
+
+        // Note: this can be undefined when the peer dependency isn't provided at all
+        const resolvedDescriptor = root.dependencies.get(ident.identHash);
+
+        if (typeof resolvedDescriptor !== `undefined`) {
+          const peerResolution = getPackageFromDescriptor(resolvedDescriptor);
+          const peerVersion = peerResolution.version ?? `0.0.0`;
+
+          const ranges = new Set<string>();
+          for (const linkHash of linkHashes) {
+            const link = allPackages.get(linkHash);
+            if (typeof link === `undefined`)
+              throw new Error(`Assertion failed: Expected the link to be registered`);
+
+            const peerDependency = link.peerDependencies.get(ident.identHash);
+            if (typeof peerDependency === `undefined`)
+              throw new Error(`Assertion failed: Expected the ident to be registered`);
+
+            ranges.add(peerDependency.range);
+          }
+
+          const satisfiesAll = [...ranges].every(range => {
+            return semverUtils.satisfiesWithPrereleases(peerVersion, range);
+          });
+
+          if (!satisfiesAll) {
+            warnings.push({
+              type: WarningType.NotCompatible,
+              subject: dependent,
+              requested: ident,
+              requester: root,
+              version: peerVersion,
+              hash,
+              requirementCount: linkHashes.length,
+            });
+          }
+        } else {
+          const peerDependencyMeta = root.peerDependenciesMeta.get(identStr);
+
+          if (!peerDependencyMeta?.optional) {
+            warnings.push({
+              type: WarningType.NotProvided,
+              subject: dependent,
+              requested: ident,
+              requester: root,
+              hash,
+            });
+          }
+        }
+      }
+    }
+  }
+
+  const warningSortCriterias: Array<((warning: Warning) => string)> = [
+    warning => structUtils.prettyLocatorNoColors(warning.subject),
+    warning => structUtils.stringifyIdent(warning.requested),
+    warning => `${warning.type}`,
+  ];
+
+  for (const warning of miscUtils.sortMap(warnings, warningSortCriterias)) {
+    switch (warning.type) {
+      case WarningType.NotProvided: {
+        report?.reportWarning(MessageName.MISSING_PEER_DEPENDENCY, `${
+          structUtils.prettyLocator(project.configuration, warning.subject)
+        } doesn't provide ${
+          structUtils.prettyIdent(project.configuration, warning.requested)
+        } (${
+          formatUtils.pretty(project.configuration, warning.hash, formatUtils.Type.CODE)
+        }), requested by ${
+          structUtils.prettyIdent(project.configuration, warning.requester)
+        }`);
+      } break;
+
+      case WarningType.NotCompatible: {
+        const andDescendants = warning.requirementCount > 1
```

Comment:
My previous implementation hid the "and descendants" part when only the root requester wasn't satisfied and all of the others were. Do you think we should add that part back?

Reply:
For now I prefer to keep that simple-ish. Not sure the added value meets the bar enough to be worth the extra complexity cost.